### PR TITLE
Feature/studio add split vertical

### LINKED ISSUE #408: studio: add split template vertical
### Description

In #372 a new option `vertical` was added to the `split` template. Per default this slide is still split horizontally ("two columns") with this new options it is now possible to split it vertically ("two rows").

The goal of this feature is to unleash this option in the editor ("studio")

### UX

We are not going to add an option which would allow such a slide to be transformed dynamically. I think that the implementation and the UX are going to easier if we "just" allow two different types of slide in the slide selector (see screenshot)

### Implementation

- In `slide.tsx` refactor enum `SlideTemplate` respectively `SPLIT` to `SPLIT_HORIZONTALLY` and then add a new value `SPLIT_VERTICALLY` **UPDATE: That was a bad idea, my bad**

- In `app-create-slide.tsx` add the new template

- Pass the `vertical` property till `create-slides.utils.tsx` to create effectively the slides. **PING ME ABOUT THIS** I've got something in development about that in an ongoing branch

- Add `vertical?: boolean` to `SlideAttributes` in `slide.tsx`

- In `deck-events.handler.tsx` method `getSlideAttributes` add the matching for vertical ("slide html element attribute vertical -> slide attributes object vertical variable")

- In `parse-slides.utils.tsx` method `parseSlideElement` do the contrary, object variable to html element attribute

It might be enough but it might also need some extra styling, notably if a piece of code would be added in such a vertical slide

### Screenshot

Capture d’écran 2019-10-13 à 13 35 36

### PR DESCRIPTION
**Pull request by timrodz**
Closes #408 😄

### CODE REVIEW

**Comment by peterpeterparker** [comment]
@timrodz so cool thx for the PR 👍 

I've to catch a train soon, will review this during my travel 😉

**Comment by timrodz** [comment]
> @timrodz so cool thx for the PR 👍
> 
> I've to catch a train soon, will review this during my travel 😉

Safe travels! Time to sleep in New Zealand, so I'll be back in ~10 hours. Tschüss!

File: studio/src/app/handlers/editor/events/deck/deck-events.handler.tsx [hunk 1]
```diff
@@ -542,6 +542,13 @@ export class DeckEventsHandler {
             if ((slide as any).imgAlt) {
                 attributes.imgAlt = (slide as any).imgAlt;
             }
+            
+            if (slide.getAttribute('vertical') === 'true') {
```
Comment (peterpeterparker) [inline studio/src/app/handlers/editor/events/deck/deck-events.handler.tsx]: I would rather like that you test here the type of node and also, if not provided, then clean it from the db (`firebase.firestore.FieldValue.delete()`).

You could have a look to method `getSlideAttributesQRCode` and adapt for `split`

File: studio/src/app/models/data/slide.tsx [hunk 2]
```diff
@@ -1,7 +1,8 @@
 export enum SlideTemplate {
     TITLE = 'title',
     CONTENT = 'content',
-    SPLIT = 'split',
+    SPLIT_HORIZONTALLY = 'split',
```
Comment (peterpeterparker) [inline studio/src/app/models/data/slide.tsx]: Actually that was a bad idea from my side, sorry for that. We should stick to only `split` aka no refactoring.

File: studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx [hunk 3]
```diff
@@ -198,12 +198,18 @@ export class AppCreateSlide {
                         <p slot="content">Content</p>
                     </deckgo-slide-content>
                 </div>
-                <div class="item" custom-tappable onClick={() => this.addSlide(SlideTemplate.SPLIT)}>
+                <div class="item" custom-tappable onClick={() => this.addSlide(SlideTemplate.SPLIT_HORIZONTALLY)}>
```
Comment (peterpeterparker) [inline studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx]: Here we stick to `SlideTemplate.SPLIT`

File: studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx [hunk 4]
```diff
@@ -198,12 +198,18 @@ export class AppCreateSlide {
...
                     <deckgo-slide-split class="showcase">
                         <p slot="start">Content</p>
                         <p slot="end">Content</p>
                     </deckgo-slide-split>
                 </div>
+                <div class="item" custom-tappable onClick={() => this.addSlide(SlideTemplate.SPLIT_VERTICALLY)}>
```
Comment (peterpeterparker) [inline studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx]: Here we use `SlideTemplate.SPLIT` and create a specific method 

````
private async addSlide(template: SlideTemplate, deck?: Deck, attributes: SlideAttributes = undefined) {
``

then we extend `create-slides.utils.tsx` method `createSlideSplit()` with the attributes too

for inspiration, you could have a look to method `createSlideChart` in same class

File: studio/src/app/utils/editor/create-slides.utils.tsx [hunk 5]
```diff
@@ -25,8 +25,10 @@ export class CreateSlidesUtils {
                 resolve(await this.createSlideTitle());
             } else if (template === SlideTemplate.CONTENT) {
                 resolve(await this.createSlideContent());
-            } else if (template === SlideTemplate.SPLIT) {
+            } else if (template === SlideTemplate.SPLIT_HORIZONTALLY) {
```
Comment (peterpeterparker) [inline studio/src/app/utils/editor/create-slides.utils.tsx]: As we stick to `SPLIT`, we don't need this anymore

File: studio/src/app/utils/editor/create-slides.utils.tsx [hunk 6]
```diff
@@ -83,7 +85,7 @@ export class CreateSlidesUtils {
         });
     }
 
-    private static createSlideSplit(): Promise<JSX.IntrinsicElements> {
+    private static createSlideSplit(vertical = false): Promise<JSX.IntrinsicElements> {
```
Comment (peterpeterparker) [inline studio/src/app/utils/editor/create-slides.utils.tsx]: See my above comment, `vertical = false` should be replaced by `attributes: SlideAttributes = undefined`

File: studio/src/app/utils/editor/create-slides.utils.tsx [hunk 7]
```diff
@@ -94,7 +96,7 @@ export class CreateSlidesUtils {
 
             const end = <section slot="end"></section>;
 
-            const slide: JSX.IntrinsicElements = <deckgo-slide-split key={uuid()}>
+            const slide: JSX.IntrinsicElements = <deckgo-slide-split vertical={vertical} key={uuid()}>
```
Comment (peterpeterparker) [inline studio/src/app/utils/editor/create-slides.utils.tsx]: Should be replaced by 

```
const slide: JSX.IntrinsicElements = <deckgo-slide-split key={uuid()} {...attributes}>
``

**Comment by peterpeterparker** [comment]
@timrodz I made a mistake in my description and I'm really sorry about this. We should not refactor `SlideTemplate`, it's important that we stick to type `SPLIT` otherwise we will face problems, notably at publishing time

I've added several comments in the PR to describe that. Let me know if it's clear or not and if you could do the modifications.

If yes, again sorry, but that's super cool
If no, no worries, I'll merge the PR in a branch and will do the modifications by myself

Let me know what solution you rather like

**Comment by timrodz** [comment]
Great suggestions @peterpeterparker — Will try to action these once I get back from work.

**Comment by peterpeterparker** [comment]
Awesome!

File: studio/src/app/handlers/editor/events/deck/deck-events.handler.tsx (hunk 1 above)
Comment (timrodz) [reply studio/src/app/handlers/editor/events/deck/deck-events.handler.tsx]: How about this?

```tsx
if ((slide as any).vertical) {
    attributes.vertical = true;
} else if (cleanFields) {
    // @ts-ignore
    attributes.vertical = firebase.firestore.FieldValue.delete();
}
```

Comment (peterpeterparker) [reply studio/src/app/handlers/editor/events/deck/deck-events.handler.tsx]: Sure it works too

File: studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx (hunk 4 above)
Comment (timrodz) [reply studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx]: There already exists a method `addSlide`, should I just add the extra parameter?

File: studio/src/app/utils/editor/create-slides.utils.tsx (hunk 7 above)
Comment (timrodz) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: I get the following error log:
```log
Type '{ style?: string; src?: string; customBackground?: string; imgSrc?: string; imgAlt?: string; content?: string; customQRCode?: boolean; type?: SlideChartType; innerRadius?: number; animation?: boolean; datePattern?: string; ... 7 more ...; key: string; }' is not assignable to type 'DeckgoSlideSplit'.
  Types of property ''customBackground'' are incompatible.
    Type 'string' is not assignable to type 'boolean'.
```

File: studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx (hunk 4 above)
Comment (peterpeterparker) [reply studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx]: no need I think for the moment as it won't be use. we could always do that later

File: studio/src/app/utils/editor/create-slides.utils.tsx (hunk 7 above)
Comment (peterpeterparker) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: Add `// @ts-ignore` before it

```
// @ts-ignore
const slide: JSX.IntrinsicElements =
```

For DB I use boolean but attributes should be string. I could have two types but doing so is fine right now.

File: studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx (hunk 4 above)
Comment (timrodz) [reply studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx]: Alright, I'll make a new method then.

Comment (peterpeterparker) [reply studio/src/app/popovers/editor/app-create-slide/app-create-slide.tsx]: Coolio, thx

File: studio/src/app/utils/editor/create-slides.utils.tsx (hunk 7 above)
Comment (timrodz) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: Problem is here:
```tsx
const slide: JSX.IntrinsicElements = <deckgo-slide-split key={uuid()} {...attributes}>
    {start}
    {end}
</deckgo-slide-split>;
```

Comment (peterpeterparker) [reply studio/src/app/utils/editor/create-slides.utils.tsx]:
```
// @ts-ignore
const slide: JSX.IntrinsicElements = <deckgo-slide-split key={uuid()} {...attributes}>
    {start}
    {end}
</deckgo-slide-split>;
```

doesn't work?

Comment (timrodz) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: Nevermind, it does work now. Wasn't receiving your messages. Only now both of them are rendering as horizontal.

Comment (peterpeterparker) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: Did you pass the `{vertical: true}` in the calling method?

Comment (timrodz) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: `app-create-slide.tsx`:
```tsx
<div class="item" custom-tappable onClick={() => this.addSlideSplit(SlideTemplate.SPLIT)}>
    <deckgo-slide-split class="showcase">
        <p slot="start">Content</p>
        <p slot="end">Content</p>
    </deckgo-slide-split>
</div>
<div class="item" custom-tappable onClick={() => this.addSlideSplit(SlideTemplate.SPLIT, null, {vertical: true})}>
    <deckgo-slide-split vertical={true} class="showcase">
        <p slot="start">Content</p>
        <p slot="end">Content</p>
    </deckgo-slide-split>
</div>
```

Comment (peterpeterparker) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: And `addSlideSplit`?

Could you commit your modifications in your branch so I could have a look?

Comment (timrodz) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: Sure, can do. 

```tsx
private async addSlideSplit(template: SlideTemplate, deck?: Deck, attributes: SlideAttributes = undefined) {
    const slide: JSX.IntrinsicElements = await CreateSlidesUtils.createSlide(template, deck, this.user);
    await this.closePopover(template, slide, attributes);
}
```

Comment (peterpeterparker) [reply studio/src/app/utils/editor/create-slides.utils.tsx]: 1. in `create-slides.utils.tsx` make `createSlideSplit` public

2. in `app-create-slide.tsx` modify method like the following

```
private async addSlideSplit(template: SlideTemplate, attributes: SlideAttributes = undefined) {
        const slide: JSX.IntrinsicElements = await CreateSlidesUtils.createSlideSplit(attributes);
        await this.closePopover(template, slide, attributes);
    }
```

 and of course then adapt `onClick` call

**Comment by peterpeterparker** [comment]
Super! all good for you @timrodz, I could merge it ?

**Comment by timrodz** [comment]
Yep, go for it @peterpeterparker!

**Comment by peterpeterparker** [comment]
Super duper awesome, thx a lot for the PR 🙏

**Pull request merged**